### PR TITLE
Added complete support for pathlib

### LINKED ISSUE #1616: Support for pathlib
### Summary

We should add support for Python's [pathlib](https://docs.python.org/3/library/pathlib.html) module.

### Rationale

Users may want to pass in a `Path` object instead of a string.

### Implementation

The type hints will change from `str` to `Union[str, bytes, os.PathLike]` according to [PEP 519](https://peps.python.org/pep-0519/#provide-specific-type-hinting-support).

### Alternatives

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by pioneerHitesh**
Added ``pathlib`` (path object) support for the entire ``dataset module``

Closes #1616

### CODE REVIEW

**Comment by pioneerHitesh** [comment]
> @microsoft-github-policy-service agree

@microsoft-github-policy-service agree

File: torchgeo/datasets/agb_live_woody_density.py [hunk 1]
```diff
@@ -59,7 +59,7 @@ class AbovegroundLiveWoodyBiomassDensity(RasterDataset):
 
     def __init__(
         self,
-        paths: Union[str, Iterable[str]] = "data",
+        paths: Union[str, Iterable[str],os.PathLike] = "data",
```
Comment (adamjstewart) [inline torchgeo/datasets/agb_live_woody_density.py:62]: Can also be iterable[pathlike]

File: torchgeo/datasets/agb_live_woody_density.py [hunk 2]
```diff
@@ -115,6 +115,10 @@ def _verify(self) -> None:
 
     def _download(self) -> None:
         """Download the dataset."""
+
+        if(isinstance(self.paths,os.PathLike)):
+            self.paths = str(self.paths)
```
Comment (adamjstewart) [inline torchgeo/datasets/agb_live_woody_density.py:120]: Is it possible to make them work if we keep them as pathlib instead of casting them?

File: torchgeo/datasets/geo.py [hunk 3]
```diff
@@ -289,6 +289,8 @@ def files(self) -> set[str]:
         # Make iterable
         if isinstance(self.paths, str):
             paths: Iterable[str] = [self.paths]
+        elif isinstance(self.paths, os.PathLike):
+            paths: Iterable[os.PathLike] = [self.paths]
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:293]: Can probably keep this to a single if-statement. Either check both or check "not iterable"

File: torchgeo/datasets/agb_live_woody_density.py (hunk 2 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/agb_live_woody_density.py:120]: Unfortunately, it isn't possible to keep it as pathlib because of the method calling `` download_url(self.url, self.paths, self.base_filename)`` which is a torchvision method and only accepts strings.

File: torchgeo/datasets/geo.py (hunk 3 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/geo.py:293]: Yes, you are correct sir, will implement this.

**Comment by adamjstewart** [comment]
@pioneerHitesh can you open a separate issue for the download problems? Maybe the URL changed.

File: torchgeo/datasets/geo.py [hunk 4]
```diff
@@ -72,7 +72,7 @@ class GeoDataset(Dataset[dict[str, Any]], abc.ABC):
        dataset = landsat7 | landsat8
     """
 
-    paths: Union[str, Iterable[str]]
+    paths: Union[str, Iterable[str], os.PathLike, Iterable[os.PathLike]]
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:75]: Let's add a type alias to make things simpler. In `torchgeo/datasets/utils.py`, you can add:
```python
import os
from typing import Union, TypeAlias

Path: TypeAlias = Union[str, bytes, os.PathLike[str], os.PathLike[bytes]]
```
Then in all other files you can simply replace `str` with `Path` where needed. This will be less code duplication.

Comment (pioneerHitesh) [reply torchgeo/datasets/geo.py:75]: Thank you for your feedback, I will work towards implementing this.

File: torchgeo/datasets/utils.py [hunk 5]
```diff
@@ -17,15 +17,23 @@
...
 from torchvision.datasets.utils import check_integrity, download_url
 from torchvision.utils import draw_segmentation_masks
 
+Path: TypeAlias = Union[
+    str,
+    bytes,
+    os.PathLike[str],
+    os.PathLike[bytes],
+    Iterable[os.PathLike[str]],
+    Iterable[os.PathLike[bytes]],
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py:35]: Iterables shouldn't be included, most of the time they aren't supported

Comment (pioneerHitesh) [reply torchgeo/datasets/utils.py:35]: oh! Thank you sir

Comment (pioneerHitesh) [reply torchgeo/datasets/utils.py:35]: Sir you have mentioned to add them in this [comment](https://github.com/microsoft/torchgeo/pull/1678#discussion_r1363698094)?

File: torchgeo/datasets/agb_live_woody_density.py [hunk 6]
```diff
@@ -59,7 +58,7 @@ class AbovegroundLiveWoodyBiomassDensity(RasterDataset):
 
     def __init__(
         self,
-        paths: Union[str, Iterable[str]] = "data",
+        paths: Path = "data",
```
Comment (adamjstewart) [inline torchgeo/datasets/agb_live_woody_density.py:61]: Path or Iterable[Path]

File: torchgeo/datasets/agb_live_woody_density.py [hunk 7]
```diff
@@ -5,15 +5,14 @@
 
 import json
 import os
-from collections.abc import Iterable
```
Comment (adamjstewart) [inline torchgeo/datasets/agb_live_woody_density.py:8]: `typing.Iterable` is deprecated, should use `collections.abc.Iterable` instead: https://docs.python.org/3/library/typing.html#typing.Iterable

File: torchgeo/datasets/agb_live_woody_density.py [hunk 8]
```diff
@@ -105,7 +104,7 @@ def _verify(self) -> None:
         # Check if the user requested to download the dataset
         if not self.download:
             raise RuntimeError(
-                f"Dataset not found in `root={self.paths}` and `download=False`, "
+                f"Dataset not found in `root={str(self.paths)}` and `download=False`, "
```
Comment (adamjstewart) [inline torchgeo/datasets/agb_live_woody_density.py:107]: This should automatically be cast to a string during formatting, were you seeing an error?

File: torchgeo/datasets/geo.py [hunk 9]
```diff
@@ -748,7 +756,7 @@ class NonGeoClassificationDataset(NonGeoDataset, ImageFolder):  # type: ignore[m
 
     def __init__(
         self,
-        root: str = "data",
+        root: Union[Path, Iterable[Path]] = "data",
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:759]: Iterables aren't supported here, only Paths

File: torchgeo/datasets/utils.py [hunk 10]
```diff
@@ -17,15 +17,17 @@
...
 
 import numpy as np
 import rasterio
 import torch
 from torch import Tensor
 from torchvision.datasets.utils import check_integrity, download_url
 from torchvision.utils import draw_segmentation_masks
+from typing_extensions import TypeAlias
 
+Path: TypeAlias = Union[str, bytes, os.PathLike[str], os.PathLike[bytes]]
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py:30]: This should come after `__all__`

File: torchgeo/datasets/utils.py [hunk 11]
```diff
@@ -737,3 +739,15 @@ def percentile_normalization(
         (img - lower_percentile) / (upper_percentile - lower_percentile + 1e-5), 0, 1
     )
     return img_normalized
+
+
+def check_instance_type(objects: Union[Path, Iterable[Path]]) -> bool:
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py:744]: Why is this needed?

File: torchgeo/datasets/agb_live_woody_density.py (hunk 8 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/agb_live_woody_density.py:107]: Yes , The error occurs specifically for byte. ``error: On Python 3 '{}'.format(b'abc') produces "b'abc'", not 'abc'; use '{!r}'.format(b'abc') if this is desired behavior when using formatted string literals ``.

File: torchgeo/datasets/utils.py (hunk 11 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/utils.py:744]: This function help to check for parameterized generic types such as ``PathLike[bytes]`` , the original isinstance fails to check parameterized generic types. Here is the the error ``Second argument to "isinstance" must be a class or tuple of classes TypeVar or generic type with type arguments not allowed``.

File: torchgeo/datasets/agb_live_woody_density.py (hunk 8 above)
Comment (adamjstewart) [reply torchgeo/datasets/agb_live_woody_density.py:108]: How do I reproduce this error? It works fine for me:
```pycon
> python3 -Werror
Python 3.11.6 (main, Oct 11 2023, 22:49:14) [Clang 15.0.0 (clang-1500.0.40.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> foo = b"foo"
>>> f"{foo} bar"
"b'foo' bar"
```

File: torchgeo/datasets/geo.py [hunk 12]
```diff
@@ -574,7 +581,7 @@ class VectorDataset(GeoDataset):
 
     def __init__(
         self,
-        paths: Union[str, Iterable[str]] = "data",
+        paths: Path = "data",
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:584]:
```suggestion
        paths: Union[Path, Iterable[Path]] = "data",
```

File: torchgeo/datasets/agb_live_woody_density.py (hunk 8 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/agb_live_woody_density.py:108]: mypy detects this an error. Here is the code and output
code:
``mypy torchgeo/datasets/agb_live_woody_density.py``
output
``torchgeo/datasets/agb_live_woody_density.py:108: error: If x = b'abc' then f"{x}" or "{}".format(x) produces "b'abc'", not "abc". If this is desired behavior, use f"{x!r}" or "{!r}".format(x). Otherwise, decode the bytes  [str-bytes-safe]
Found 3 errors in 2 files (checked 1 source file)``

File: torchgeo/datasets/utils.py (hunk 11 above)
Comment (adamjstewart) [reply torchgeo/datasets/utils.py]: It doesn't have a docstring, but I'm guessing this function returns True if the input is a pathlib.Path or an iterable of pathlib.Path? But then you always cast to str, which doesn't make sense for iterables. Also, `str(object)` will always be an instance of str, so that check doesn't make sense. Can't we just check if it's an instance of `pathlib.Path`?

**Comment by adamjstewart** [comment]
Some of the type ignores are for files that have not yet been updated (this is fine), but some of the type ignores are for files that have been updated (this makes me think there's a bug in the type hints).

**Comment by pioneerHitesh** [comment]
> Some of the type ignores are for files that have not yet been updated (this is fine), but some of the type ignores are for files that have been updated (this makes me think there's a bug in the type hints).
Will investigate this and get back to you.

File: torchgeo/datasets/agb_live_woody_density.py (hunk 8 above)
Comment (adamjstewart) [reply torchgeo/datasets/agb_live_woody_density.py:108]: Can we use `{self.paths!r}` or `{self.paths!s}` instead of `{str(self.paths)}`?

File: torchgeo/datasets/utils.py (hunk 11 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/utils.py]: if we try to check if the parameter is an instance of pathlib.Path using ``isinstance`` method , Here are the results.

 Code
``isinstance(self.paths, Path)``
Errors:
```plaintext
torchgeo/datasets/geo.py:297: error: Parameterized generics cannot be used with class or instance checks  [misc]
torchgeo/datasets/geo.py:297: error: Argument 2 to "isinstance" has incompatible type "<typing special form>"; 
expected "_ClassInfo"  [arg-type]
```

The cast to str is required to find out if the path's type is str or bytes. Here is an example 
```from os import PathLike
from pathlib import Path
dir_path = Path("/home")
print(isinstance(dir_path,str))  # This will return false 
print(hasattr(dir_path,'decode'))
```

As you have mentioned str(object) will be an instance of str , so that check does not make sense and will incorrectly convert objects to string. Here is what i mean.

```from pathlib import Path
dir_path = Path.cwd()  # return type is <class 'pathlib.PosixPath'>
dir_path = Path("/home")
print(isinstance(str(dir_path), str))  # returns true
```

A better way of checking for strings would be using the ``hasattr(str_to_check, "encode")``.

Comment (adamjstewart) [reply torchgeo/datasets/utils.py]: I'm guessing that error is because you're using:
```
from pathlib import Path
```
AND
```
from .utils import Path
```
and the names conflict. Not sure if we should use a different name or just use `pathlib.Path`.

Do we need to check whether or not it's bytes? Can we treat str and bytes identically?

Comment (pioneerHitesh) [reply torchgeo/datasets/utils.py]: No, sir both of them have been tested in seperate files. `from pathlib import Path` and `from .utils import Path` are not imported simultaneously.  We can skip the test if desired so.

Comment (adamjstewart) [reply torchgeo/datasets/utils.py]: I have the following file:
```python
from pathlib import Path

p = Path('.')

if isinstance(p, Path):
    print("Path")
else:
    print("str or bytes")
```
`mypy --strict` passes on this file. Anything I'm missing here?

I don't think you need to check `Iterable[Path]`, just `Path`.

Comment (pioneerHitesh) [reply torchgeo/datasets/utils.py]: It throws the same error when i try to run the below code.
```import os
from typing import Union
from typing_extensions import TypeAlias

Path: TypeAlias = Union[str, bytes, os.PathLike[bytes], os.PathLike[str]]

testPath: Path = os.getcwd()

print(isinstance(testPath, Path))```

Comment (adamjstewart) [reply torchgeo/datasets/utils.py]: Yes, because `torchgeo.datasets.utils.Path` and `pathlib.Path` are completely different things.

File: torchgeo/datasets/agb_live_woody_density.py (hunk 8 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/agb_live_woody_density.py:108]: Yes, I have implemented this fix.

File: tests/datasets/test_advance.py [hunk 13]
```diff
@@ -15,6 +14,7 @@
 
 import torchgeo.datasets.utils
 from torchgeo.datasets import ADVANCE
+from torchgeo.datasets.utils import Path
```
Comment (adamjstewart) [inline tests/datasets/test_advance.py:17]: This is incorrect, the correct type is `pathlib.Path`

Comment (pioneerHitesh) [reply tests/datasets/test_advance.py:17]: Have replaced it with Path , so that it stays consistent with the codebase.

Comment (adamjstewart) [reply tests/datasets/test_advance.py:17]: `torchgeo.datasets.utils.Path` and `pathlib.Path` are two different things. The latter is a subset of the former. We should be as specific as possible.

Comment (pioneerHitesh) [reply tests/datasets/test_advance.py:17]: @adamjstewart ok will revert these changes.

**Comment by adamjstewart** [comment]
Can you replace all instances of `str(tmp_path)` with `tmp_path` in our tests? Should be possible with a 1-line sed command. Then we can actually test to ensure that `pathlib.Path` works correctly.

**Comment by pioneerHitesh** [comment]
> Can you replace all instances of `str(tmp_path)` with `tmp_path` in our tests? Should be possible with a 1-line sed command. Then we can actually test to ensure that `pathlib.Path` works correctly.

@adamjstewart  Unfortunately, it not completely possible to replace ``str(tmp_path)`` because ``download_url()`` method does not support path but i am trying to reduce ``str(tmp_path)``.

**Comment by pioneerHitesh** [comment]
@adamjstewart any additional changes ?

File: tests/datasets/test_eudem.py [hunk 14]
```diff
@@ -33,22 +34,22 @@ def test_getitem(self, dataset: EUDEM) -> None:
         assert isinstance(x["mask"], torch.Tensor)
 
     def test_extracted_already(self, dataset: EUDEM) -> None:
-        assert isinstance(dataset.paths, str)
-        zipfile = os.path.join(dataset.paths, "eu_dem_v11_E30N10.zip")
-        shutil.unpack_archive(zipfile, dataset.paths, "zip")
+        assert check_instance_type(dataset.paths)
```
Comment (adamjstewart) [inline tests/datasets/test_eudem.py:37]: I think we can just assert that this is a `pathlib.Path`

File: tests/datasets/test_landcoverai.py [hunk 15]
```diff
@@ -44,13 +44,13 @@ def test_already_extracted(self, dataset: LandCoverAIGeo) -> None:
 
     def test_already_downloaded(self, monkeypatch: MonkeyPatch, tmp_path: Path) -> None:
         url = os.path.join("tests", "data", "landcoverai", "landcover.ai.v1.zip")
-        root = str(tmp_path)
+        root = tmp_path
         shutil.copy(url, root)
-        LandCoverAIGeo(root)
+        LandCoverAIGeo(str(root))
```
Comment (adamjstewart) [inline tests/datasets/test_landcoverai.py:49]: Why does this need to be converted to a string?

File: tests/datasets/test_mapinwild.py [hunk 16]
```diff
@@ -98,20 +98,20 @@ def test_invalid_split(self) -> None:
...
 
     def test_downloaded_not_extracted(self, tmp_path: Path) -> None:
         pathname = os.path.join("tests", "data", "mapinwild", "*", "*")
         pathname_glob = glob.glob(pathname)
-        root = str(tmp_path)
+        root = tmp_path
         for zipfile in pathname_glob:
             shutil.copy(zipfile, root)
-        MapInWild(root, download=False)
+        MapInWild(str(root), download=False)
```
Comment (adamjstewart) [inline tests/datasets/test_mapinwild.py:109]: Why str?

File: torchgeo/datasets/agb_live_woody_density.py [hunk 17]
```diff
@@ -105,7 +105,7 @@ def _verify(self) -> None:
         # Check if the user requested to download the dataset
         if not self.download:
             raise RuntimeError(
-                f"Dataset not found in `root={self.paths}` and `download=False`, "
+                f"Dataset not found in `root={self.paths!r}` and `download=False`, "
```
Comment (adamjstewart) [inline torchgeo/datasets/agb_live_woody_density.py:108]:
```suggestion
                f"Dataset not found in `paths={self.paths!r}` and `download=False`, "
```
We forgot to update this in the root -> paths PR, may need to be fixed in multiple files

File: torchgeo/datasets/astergdem.py [hunk 18]
```diff
@@ -88,7 +90,7 @@ def _verify(self) -> None:
             return
 
         raise RuntimeError(
-            f"Dataset not found in `root={self.paths}` "
+            f"Dataset not found in `root={self.paths!r}` "
```
Comment (adamjstewart) [inline torchgeo/datasets/astergdem.py:93]:
```suggestion
            f"Dataset not found in `paths={self.paths!r}` "
```

File: torchgeo/datasets/benin_cashews.py [hunk 19]
```diff
@@ -202,7 +207,7 @@ def __init__(
         """
         self._validate_bands(bands)
 
-        self.root = os.path.expanduser(root)
+        self.root = str(os.path.expanduser(root))
```
Comment (adamjstewart) [inline torchgeo/datasets/benin_cashews.py:210]: For uniformity we prob shouldn't use `os.path.expanduser` in any dataset

File: torchgeo/datasets/cbf.py [hunk 20]
```diff
@@ -108,9 +113,9 @@ def _check_integrity(self) -> bool:
         Returns:
             True if dataset files are found and/or MD5s match, else False
         """
-        assert isinstance(self.paths, str)
+        assert check_instance_type(self.paths)
```
Comment (adamjstewart) [inline torchgeo/datasets/cbf.py:116]: Seems like this dataset doesn't actually support iterables. If we change the type hints to just `Path` does mypy complain?

File: torchgeo/datasets/cdl.py [hunk 21]
```diff
@@ -314,7 +314,7 @@ def _verify(self) -> None:
         # Check if the user requested to download the dataset
         if not self.download:
             raise RuntimeError(
-                f"Dataset not found in `root={self.paths}` and `download=False`, "
+                f"Dataset not found in `root={self.paths!r}` and `download=False`, "
```
Comment (adamjstewart) [inline torchgeo/datasets/cdl.py:317]:
```suggestion
                f"Dataset not found in `paths={self.paths!r}` and `download=False`, "
```

File: torchgeo/datasets/cms_mangrove_canopy.py [hunk 22]
```diff
@@ -235,24 +236,24 @@ def _verify(self) -> None:
...
+        pathname = os.path.join(str(self.paths), self.zipfile)
         if os.path.exists(pathname):
             if self.checksum and not check_integrity(pathname, self.md5):
                 raise RuntimeError("Dataset found, but corrupted.")
             self._extract()
             return
 
         raise RuntimeError(
-            f"Dataset not found in `root={self.paths}` "
+            f"Dataset not found in `root={self.paths!r}` "
```
Comment (adamjstewart) [inline torchgeo/datasets/cms_mangrove_canopy.py:248]:
```suggestion
            f"Dataset not found in `paths={self.paths!r}` "
```

File: torchgeo/datasets/geo.py [hunk 23]
```diff
@@ -287,16 +294,17 @@ def files(self) -> set[str]:
...
         """
         # Make iterable
-        if isinstance(self.paths, str):
-            paths: Iterable[str] = [self.paths]
+        paths: Iterable[Path]
+        if check_instance_type(self.paths):
+            paths = [self.paths]  # type: ignore[list-item]
         else:
-            paths = self.paths
+            paths = self.paths  # type: ignore[assignment]
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:301]: Something is definitely wrong here...

File: torchgeo/datasets/geo.py [hunk 24]
```diff
@@ -392,7 +400,7 @@ def __init__(
         i = 0
         filename_regex = re.compile(self.filename_regex, re.VERBOSE)
         for filepath in self.files:
-            match = re.match(filename_regex, os.path.basename(filepath))
+            match = re.match(filename_regex, os.path.basename(str(filepath)))
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:403]: I don't think this needs to be cast to a str, right?

File: torchgeo/datasets/geo.py [hunk 25]
```diff
@@ -432,7 +440,6 @@ def __init__(
             if self.bands:
                 msg += f" with `bands={self.bands}`"
             raise FileNotFoundError(msg)
-
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:435]: Can you add this blank line back?

File: torchgeo/datasets/geo.py [hunk 26]
```diff
@@ -629,7 +636,8 @@ def __init__(
                 i += 1
 
         if i == 0:
-            msg = f"No {self.__class__.__name__} data was found in `root='{paths}'`"
+            str_path = str(paths)
+            msg = f"No {self.__class__.__name__} data was found in `root='{str_path}'`"
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:640]:
```suggestion
            msg = f"No {self.__class__.__name__} data was found in `paths={paths!r}`"
```

File: torchgeo/datasets/inaturalist.py [hunk 27]
```diff
@@ -44,9 +44,9 @@ def __init__(self, root: str = "data") -> None:
         """
         super().__init__()
 
-        self.root = root
+        self.root = str(root)
 
-        files = glob.glob(os.path.join(root, "**.csv"))
+        files = glob.glob(os.path.join(self.root, "**.csv"))
```
Comment (adamjstewart) [inline torchgeo/datasets/inaturalist.py:49]: This doesn't need to be a str, right?

File: torchgeo/datasets/nlcd.py [hunk 28]
```diff
@@ -209,7 +209,7 @@ def _verify(self) -> None:
         # Check if the user requested to download the dataset
         if not self.download:
             raise RuntimeError(
-                f"Dataset not found in `root={self.paths}` and `download=False`, "
+                f"Dataset not found in `root={str(self.paths)}` and `download=False`, "
```
Comment (adamjstewart) [inline torchgeo/datasets/nlcd.py:212]:
```suggestion
                f"Dataset not found in `paths={self.paths!r}` and `download=False`, "
```

File: torchgeo/datasets/openbuildings.py [hunk 29]
```diff
@@ -278,14 +278,14 @@ def __init__(
...
-                self.paths, feature["properties"]["tile_url"].split("/")[-1]
+                str(self.paths), feature["properties"]["tile_url"].split("/")[-1]
             )
             self.index.insert(i, coords, filepath)
             i += 1
 
         if i == 0:
             raise FileNotFoundError(
-                f"No {self.__class__.__name__} data was found in '{self.paths}'"
+                f"No {self.__class__.__name__} data was found in '{self.paths!r}'"
```
Comment (adamjstewart) [inline torchgeo/datasets/openbuildings.py:288]:
```suggestion
                f"No {self.__class__.__name__} data was found in `paths={self.paths!r}`"
```

File: torchgeo/datasets/utils.py [hunk 30]
```diff
@@ -45,6 +46,8 @@
     "percentile_normalization",
 )
 
+Path: TypeAlias = Union[str, bytes, os.PathLike[bytes], os.PathLike[str]]
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py:49]:
```suggestion
Path: TypeAlias = Union[str, bytes, os.PathLike[str], os.PathLike[bytes]]
```
Just to keep the order consistent

File: torchgeo/datasets/utils.py [hunk 31]
```diff
@@ -17,14 +17,15 @@
...
-from typing import Any, cast, overload
+from typing import Any, Union, cast, overload
 
 import numpy as np
 import rasterio
 import torch
 from torch import Tensor
 from torchvision.datasets.utils import check_integrity, download_url
 from torchvision.utils import draw_segmentation_masks
+from typing_extensions import TypeAlias
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py:28]: We don't have a dep on typing_extensions right now. Let's just remove the `TypeAlias` usage, we don't actually need it.

File: torchgeo/datasets/utils.py [hunk 32]
```diff
@@ -737,3 +740,25 @@ def percentile_normalization(
         (img - lower_percentile) / (upper_percentile - lower_percentile + 1e-5), 0, 1
     )
     return img_normalized
+
+
+def check_instance_type(paths: Path | Iterable[Path]) -> bool:
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py:745]: I don't love this function:

* The name is way too vague
* The docstring doesn't tell me when it's True and when it's False
* Do we really need this complex of a check? 

It seems like we're really just checking if `paths` is a `Path` (True) or `Iterable[Path]` (False). Can't we just do:
```python
return not isinstance(paths, Iterable)
```
? In that case, we wouldn't even need a function. Basically, if we can figure out a reasonable way to remove this function, I would be happier.

**Review by adamjstewart — COMMENTED** [review]
At the moment, it seems like 99% of the time, we accept any Path as input, then immediately convert it to a str. I think this is still useful, but it's a lot of extra work to ensure valid types and cast everything. Is torchvision's download_url the only reason we need to cast to a str? I wonder if we could just update download_url to accept any Path as input. I'm willing to submit a PR to torchvision if you think that's worth it. We could even have a wrapper in TorchGeo that just converts the path to a str and calls the torchvision version. Let me know what you think of this idea.

Also, this PR is becoming massive. Is there any way we can split this up into multiple PRs to make it easier on both of us to update/review? Maybe 1 PR to remove `os.path.expanduser`, another PR to fix `root={self.paths}` to `paths={self.paths!r}`, another PR to add support to the datasets, and another PR to update the tests? Or would that be too much work for you?

File: torchgeo/datasets/agb_live_woody_density.py (hunk 17 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/agb_live_woody_density.py:108]: Noted, will look into this.

File: torchgeo/datasets/utils.py (hunk 32 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/utils.py:745]: @adamjstewart Unfortunately ``isinstance(paths, Iterable)`` will also return ``True`` for ``bytes and str``. will have to find another way to eliminate this function.

**Comment by pioneerHitesh** [comment]
> At the moment, it seems like 99% of the time, we accept any Path as input, then immediately convert it to a str. I think this is still useful, but it's a lot of extra work to ensure valid types and cast everything. Is torchvision's download_url the only reason we need to cast to a str? I wonder if we could just update download_url to accept any Path as input. I'm willing to submit a PR to torchvision if you think that's worth it. We could even have a wrapper in TorchGeo that just converts the path to a str and calls the torchvision version. Let me know what you think of this idea.
> 
> Also, this PR is becoming massive. Is there any way we can split this up into multiple PRs to make it easier on both of us to update/review? Maybe 1 PR to remove `os.path.expanduser`, another PR to fix `root={self.paths}` to `paths={self.paths!r}`, another PR to add support to the datasets, and another PR to update the tests? Or would that be too much work for you?

In addition to the torchvision's ``download_url`` , the ``os.path.join`` also needs ``str``. If we can drop support for Path being of type ``bytes`` and ``pathlib.Path[bytes]`` then we do not need to cast to ``str``. The problem with ``bytes`` or ``pathlib.Path[bytes]`` is if the second argument passed in the ``join`` method is a string it is unable to combine the paths. It also does not support merging of ``pathlib.Path[bytes]`` and ``bytes``.  Another issue i have discovered is ``pathlib.path`` does not support ``bytes``. So we will have to drop support for atleast ``os.PathLike[bytes]`` and if feasible also ``bytes``. 

My suggestion would be that we add support for ``os.PathLike[str]`` in torchvision's ``download_url`` as that would benefit others as well and will eliminate the need for writing a wrapper function just to cast to ``str``. I concur with you regarding the PR becoming unviable to update/review and will split the PR. However we cannot split the ``datasets`` and ``tests`` separately as that would lead to failing of tests instead we could split ``datasets`` and ``tests`` by limiting the number of ``datasets`` and ``tests`` files changed in a commit .

File: tests/datasets/test_mapinwild.py (hunk 16 above)
Comment (pioneerHitesh) [reply tests/datasets/test_mapinwild.py:109]: This was probably added earlier to fix mypy error when ``MapInWild`` did not support ``pathlib.path``. Have fixed this and gone through other test files to fix similar issues.

File: tests/datasets/test_landcoverai.py (hunk 15 above)
Comment (pioneerHitesh) [reply tests/datasets/test_landcoverai.py:49]: This was probably added earlier to fix mypy error when ``LandCoverAIGeo`` did not support ``pathlib.path``. Have fixed this and gone through other test files to fix similar issues.

File: tests/datasets/test_eudem.py (hunk 14 above)
Comment (pioneerHitesh) [reply tests/datasets/test_eudem.py:37]: @adamjstewart  We may not be able to do that because we have to assert for ``str``,``bytes`` , ``os.PathLike[str]`` and
``os.PathLike[bytes]``.

File: torchgeo/datasets/geo.py (hunk 23 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/geo.py:301]: Here the issue is with mypy wrongly throwing an error. The ``check_instance_type()`` will return ``True`` if it is not a iterable 
so inner body of if loop will cast the ``paths`` to a ``list[paths]`` and if it will be ``False`` which means it is iterable, then the paths is directly assigned to ``self.paths``.
Below are the code and error

Code
``paths = [self.paths]``
Output
``` List item 0 has incompatible type "Union[bytes, PathLike[bytes], PathLike[str], Iterable[Union[str, bytes, PathLike[bytes], PathLike[str]]]]"; expected "Union[str, bytes, PathLike[bytes], PathLike[str]]"```

Code
``paths = self.paths`` 
Output
```torchgeo/datasets/geo.py:301: error: Incompatible types in assignment (expression has type "Union[bytes, PathLike[bytes], PathLike[str], Iterable[Union[str, bytes, PathLike[bytes], PathLike[str]]]]", variable has type "Iterable[Union[str, bytes, PathLike[bytes], PathLike[str]]]")  [assignment]```

File: torchgeo/datasets/cbf.py (hunk 20 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/cbf.py:116]: @adamjstewart 

Code:
``assert isinstance(self.paths, Path)``.

Mypy error:
```
Parameterized generics cannot be used with class or instance checks  [misc]
Argument 2 to "isinstance" has incompatible type "<typing special form>"; expected "_ClassInfo"  [arg-type]
```

File: torchgeo/datasets/benin_cashews.py (hunk 19 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/benin_cashews.py:210]: @adamjstewart  This is noted , will remove this later.

File: torchgeo/datasets/inaturalist.py (hunk 27 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/inaturalist.py:49]: This will need to be ``str`` as ``join`` does not support concatenation of ``bytes`` and ``str``.

**Comment by pioneerHitesh** [comment]
> > However we cannot split the datasets and tests separately as that would lead to failing of tests instead we could split datasets and tests by limiting the number of datasets and tests files changed in a commit .
> 
> We could fix the datasets in one PR and then update the tests in a second PR after the first one is merged.

Should I start implementing the changes?

File: torchgeo/datasets/geo.py (hunk 24 above)
Comment (pioneerHitesh) [reply torchgeo/datasets/geo.py:403]: This will need to be ``str`` as join does not support concatenation of ``bytes`` and ``str``.

Comment (pioneerHitesh) [reply torchgeo/datasets/geo.py:403]: Possibly ``bytes`` and ``os.PathLike`` is not supported in ``re.match``. Either both arguments in ``re.match`` should be ``str`` or ``bytes``.
[More info here](https://docs.python.org/3/library/re.html)

**Comment by adamjstewart** [comment]
Yeah, let's start with PRs for `os.path.expanduser` and `root={self.paths}`, then once those are merged a PR for `Path`.

**Comment by pioneerHitesh** [comment]
> os.path.expanduser
Creating a separate issue would be better? This PR already has 75 comments.

**Comment by adamjstewart** [comment]
We can either close this PR and start new PRs or mark this one as a draft and rebase once other PRs are merged. I have no preference.

**Comment by pioneerHitesh** [comment]
> We can either close this PR and start new PRs or mark this one as a draft and rebase once other PRs are merged. I have no preference.

I think we should convert this to draft.

**Comment by pioneerHitesh** [comment]
> Yeah, let's start with PRs for `os.path.expanduser` and `root={self.paths}`, then once those are merged a PR for `Path`.

should i rename ``root`` to ``path`` in   ``__init__``  methods?

**Comment by adamjstewart** [comment]
For NonGeoDatasets we want to keep using `root`. For GeoDatasets, we should use `paths` everywhere, including in error messages.

**Comment by pioneerHitesh** [comment]
@adamjstewart  please take a look at this [PR](https://github.com/microsoft/torchgeo/pull/1702).

**Comment by adamjstewart** [comment]
Closing in favor of #1731
